# Exploração e limpeza de dados

### Diferentes tipos de problemas da ciência de dados

- problema de regressão

Por exemplo, prever o preço de uma casa. O esperado é que fique perto do preço real

- problema de classificação

Por exemplo, prever uma resposta sim ou não para uma pergunta. O esperado é a resposta correta.

Ambos problemas são chamados de **aprendizado supervisionado**, um tipo de problema que depende de dados rotulados

### DataFrame

DataFrame é uma classe básica do pandas. Funciona como um template para uma estrutura de dados.

Exemplo de um DataFarme:

|    | ID | Sex | Name |
|----|----|-----|------|
| 1  | 001|M    | Pedro|
| 2  | 002|F    | Ana  |

Tabela: *Exemplo de DataFrame do pandas com um ímdice de linha inteiro à esquerda e um índice de coluna na forma de strings*

In [1]:
# Carregando os dados do estudo de caso em um jupyter notebook

## importar pandas

import pandas as pd

In [2]:
# importando o dataset usando o método pd.read_excel

df = pd.read_excel('C:/Users/Renato/OneDrive/github/Data_science_for_architecture/projetos_de_ciencias_de_dados_com_python/Data/default_of_credit_card_clients__courseware_version_1_21_19.xls')

In [3]:
# testando a importação do dataset via pandas
# Use o método .info() para ver todas as colunas com os tipos de dados

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 25 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   ID                          30000 non-null  object
 1   LIMIT_BAL                   30000 non-null  int64 
 2   SEX                         30000 non-null  int64 
 3   EDUCATION                   30000 non-null  int64 
 4   MARRIAGE                    30000 non-null  int64 
 5   AGE                         30000 non-null  int64 
 6   PAY_1                       30000 non-null  object
 7   PAY_2                       30000 non-null  int64 
 8   PAY_3                       30000 non-null  int64 
 9   PAY_4                       30000 non-null  int64 
 10  PAY_5                       30000 non-null  int64 
 11  PAY_6                       30000 non-null  int64 
 12  BILL_AMT1                   30000 non-null  int64 
 13  BILL_AMT2                   30000 non-null  in

### O problema da empresa

Cliente é uma empresa de cartão de crédito. O dataset inclui dados demográficos e financeiros recentes (6 últimos meses) de uma amostra de 30.000 titulares.

As linhas são rotuladas de acordo com se no mês seguinte ao período de dados histórico de seis meses um proprietário de conta ficou inadimplente, ou seja, não fez o pagamento mínimo.

### Objetivo

Desenvolver um modelo que preveja se uma conta ficará inadimplente no próximo mês.

# Etapa de exploração de dados

Algumas etapas úteis para a exploração de dados:  

- saber quantas colunas os dados contêm (características, resposta ou metadados);
- quantas linhas (amostras);
- que tipos de características existem (categóricas (sim/não/talvez) ou númericas);
- qual é a aparência dos dados segundo essas características;
- há dados faltando?
    

In [4]:
# Verificando a integridade básica dos dados

# Empregar o método .columns do DataFrame para examinar os nomes de todas as colunas.

df.columns

Index(['ID', 'LIMIT_BAL', 'SEX', 'EDUCATION', 'MARRIAGE', 'AGE', 'PAY_1',
       'PAY_2', 'PAY_3', 'PAY_4', 'PAY_5', 'PAY_6', 'BILL_AMT1', 'BILL_AMT2',
       'BILL_AMT3', 'BILL_AMT4', 'BILL_AMT5', 'BILL_AMT6', 'PAY_AMT1',
       'PAY_AMT2', 'PAY_AMT3', 'PAY_AMT4', 'PAY_AMT5', 'PAY_AMT6',
       'default payment next month'],
      dtype='object')

Como podemos ver, os nomes de todas as colunas estão listadas na saída. A coluna IDs de contas chama-se ID. As outras colunas parecem ser as **características**, com a última coluna sendo a **variável de resposta**.

Resumo das informações do dataset que nos foi dado pela clinte:

- 'ID - coluna de IDs de contas'; 
- 'LIMIT_BAL' - valor do crédito fornecido inclusive o crédito do consumidor individual e familiar (complementar); 
- 'SEX' - gênero (1=masculino; 2=feminino); 
- 'EDUCATION' - instrução (1=pós-graduação; 2=universidade; 3=ensino médio; 4=outros); 
- 'MARRIAGE' - estado civil (1=casado; 2=solteiro; 3=outros), 
- 'AGE' - idade (ano); 
- 'PAY_1'-'PAY_6' - registro de pagamentos passados.  Pagamentos mensais passados, registrados de abril a setembro, são armazenados nessas colunas;
- 'PAY_1' - representa o status de reembolso em setembro;
- 'PAY_2', - status de reembolso em agosto;
- 'PAY_3', - status de reembolso em julho;
- 'PAY_4', - status de reembolso em junho;
- 'PAY_5', - status de reembolso em maio;
- 'PAY_6', - status de reembolso em abril;

A escala de medida do status de reembolso é a seguinte: -1 = pagamento pontual; 1 = atraso de um mês no pagamento; 2 = atraso de dois meses no pagamento; e assim por diante até 8 = atraso de oito meses no pagamento; 9 = atraso de nove meses ou mais no pagamento.

- 'BILL_AMT1'-'PAY_AMT6', - valorda fatura; 
- 'BILL_AMT1' - representa o valor da fatura em setembro;
- 'BILL_AMT2' - representa o valor da fatura em agosto;
- 'BILL_AMT3' - representa o valor da fatura em julho; 
- 'BILL_AMT4' - representa o valor da fatura em junho; 
- 'BILL_AMT5' - representa o valor da fatura em maio; 
- 'BILL_AMT6' - representa o valor da fatura em abril; 
- 'PAY_AMT1'-'PAY_AMT6' - valor de pagamentos anteriores;
- 'PAY_AMT1' - representa o valor pago em setembro;
- 'PAY_AMT2' - representa o valor pago em agosto; 
- 'PAY_AMT3' - representa o valor pago em julho; 
- 'PAY_AMT4' - representa o valor pago em junho; 
- 'PAY_AMT5' - representa o valor pago em maio; 
- 'PAY_AMT6' - representa o valor pago em abril;
- 'default payment next month'

Notas: 
- não usaremos os dados de gênero para tomar decisões de solvibilidade devido a cosiderações éticas;
- valores em novos dólares taiwaneses; 

In [5]:
# Use o método .head() para ver as primeiras linhas do DataFrame

df.head()

ID  LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE PAY_1  PAY_2  \
0  798fc410-45c1      20000    2          2         1   24     2      2   
1  8a8c8f3b-8eb4     120000    2          2         2   26    -1      2   
2  85698822-43f5      90000    2          2         2   34     0      0   
3  0737c11b-be42      50000    2          2         1   37     0      0   
4  3b7f77cc-dbc0      50000    1          2         1   57    -1      0   

   PAY_3  PAY_4  ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2  \
0     -1     -1  ...          0          0          0         0       689   
1      0      0  ...       3272       3455       3261         0      1000   
2      0      0  ...      14331      14948      15549      1518      1500   
3      0      0  ...      28314      28959      29547      2000      2019   
4     -1      0  ...      20940      19146      19131      2000     36681   

   PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  default payment next month  
0         0         0         0         0                           1  
1      1000      1000         0      2000                           1  
2      1000      1000      1000      5000                           0  
3      1200      1100      1069      1000                           0  
4     10000      9000       689       679                           0  

[5 rows x 25 columns]

A coluna ID parece conter identificadores exclusivos. Para verificar se isso ocorre realmente em todo o dataset, podemos contar o número de valores exclusivos usando o método **.nunique()** da **série** (ou seja, coluna) ID. Primeiro, selecionaremos a coluna usando colchetes.

In [6]:
# descobrindo um problema de qualidade de dados
# A função pandas .nunique() retorna o número de elementos únicos no objeto.

df['ID'].nunique()

29687

Agora, vamos executar o comando **.shape** para descobrir o número de linhas do dataset

In [7]:
# descobrindo o números de linhas do dataset

df.shape

(30000, 25)

Identificamos que o número de IDs exclusivos é menor do que o número de linhas. Isso significa que o **ID não é um identificador exclusivo** para as linhas de dados.

Sabemos então que há alguma duplicação de IDs. No entanto, em que quantidades? Um único ID está sendo duplicado várias vezes? Quantos IDs estão sendo duplicados?

Podemos usar o método **.value_counts()** na série ID para começar a responder a essas perguntas. Ele é semelhante a um procedimento **group by/count** em SQL e listará os IDs exclusivos e a frequência com que ocorrem. Executaremos essa operação na próxima etapa e armazenaremos as contagens de valores em uma variável **id_counts**.

In [8]:
# armazene as contagens de valores em uma variável definida como id_counts e exiba os valores armazenados usando o método .head()

id_counts = df['ID'].value_counts()

In [9]:
id_counts.head()

6ff744d5-91f0    2
4653bc48-0141    2
5deff6b6-62ff    2
edd737fd-5b07    2
f0723a52-7018    2
Name: ID, dtype: int64

Observe que o **.head()** retorna, por padrão, as cinco primeiras linhas. Podemos especificar o número de itens a serem exibidos passando o número desejado nos parênteses, ().

In [10]:
id_counts.head(15)

6ff744d5-91f0    2
4653bc48-0141    2
5deff6b6-62ff    2
edd737fd-5b07    2
f0723a52-7018    2
e77f7765-8731    2
3d832783-1571    2
15edb406-305b    2
994330b6-eeec    2
369e565a-26a1    2
87dec940-75b7    2
e50d8395-da32    2
04c32736-903c    2
e668604e-21f6    2
4f249cbc-5e9c    2
Name: ID, dtype: int64

In [11]:
# exiba o número de entradas duplicadas agrupadas executando outra contagem de valores

id_counts.value_counts()

1    29374
2      313
Name: ID, dtype: int64

Na saída acima podemos ver que a maioria dos IDs ocorre exatamente uma única vez, como esperado. No entanto, 313 IDs ocorrem duas vezes. Logo, nenhum ID ocorre mais do que duas vezes.

De posse dessas informações, estamos prontos para começar a examinar mais detalhadamente esse problema de qualidade de dados e corrigi-lo. Criaremos máscaras booleanas para limpar melhor os dados.

# Máscaras booleanas

Para ajudar a limpar os dados do estudo de caso, usaremos o conceito de **máscara lógica**, também conhecido como **máscara booleana**.

Uma máscara lógica é uma maneira de filtrar um array, ou série (ou seja, coluna), obedecendo alguma condição. Por exemplo, podemos usar o operador "==; <; >; <=; >=" para encontrar todos os locais de um array que contêm um valor específico.

A saída dessa comparação é um array com uma série de valores  True/False, também conhecidos como valores **booleanos**. Cada elemento da saída corresponderá a um elemento da entrada, com o resultado **True** se a condição for atendida, e **False**, caso não seja.


Para ilustrar como isso funciona, usaremos **dados sintéticos** (dados criados para a exploração ou ilustração de um conceito). primeiro importaremos o pacote NumPy, para gerar números aleatórios.

In [12]:
# importando a biblioteca NumPy

import numpy as np

In [13]:
# usamos seed (semente) no gerador de números aleatórios

np.random.seed(seed=24)

Se definirmos um **seed**, teremos os mesmos resultados enre as execuções do gerador. Caso contrário, isso não é garantido. 

Essa pode ser uma opção útil se você usa números aleatórios em seu trabalho e quer ter resultados consistentes sempre que executar um notebook

In [14]:
# geração de 100 inteiros aleatórios, selecinodos entre 1 e 5 (inclusive)

random_integers = np.random.randint(low=1, high=5, size=100)

In [15]:
random_integers

array([3, 4, 1, 4, 2, 2, 2, 1, 4, 4, 1, 4, 4, 3, 4, 4, 4, 4, 4, 4, 2, 3,
       4, 4, 2, 4, 2, 4, 1, 1, 3, 1, 4, 2, 2, 1, 4, 3, 2, 2, 3, 3, 2, 3,
       4, 2, 3, 4, 1, 3, 4, 2, 2, 4, 1, 2, 3, 3, 2, 1, 2, 4, 2, 1, 4, 3,
       1, 2, 1, 3, 4, 2, 3, 4, 1, 3, 2, 1, 1, 3, 3, 4, 3, 2, 4, 1, 4, 4,
       4, 4, 1, 4, 3, 2, 4, 4, 3, 3, 2, 4])

In [16]:
# obtendo os cinco primeiros números

random_integers[:5]

array([3, 4, 1, 4, 2])

Suponhamos que quiséssemos conhecer os locais de todos os elementos de **random_integer igual a 3**. Poderíamos criar uma máscara booleana para fazer isso.

In [17]:
# criando máscara booleana para condições iguais a 3

is_equal_to_3 = random_integers == 3

In [18]:
# obtendo os cinco primeiros resutados

is_equal_to_3[:5]

array([ True, False, False, False, False])

In [19]:
# somando o resultados da máscara booleana

sum(is_equal_to_3)

22

O resultado igual a 22 faz sentido, com uma criação aleátoria igualmente provável de 5 valores possíveis, o esperado é que cada valor apareça 20 % das vezes

In [20]:
# selecionando os elementos do array original que atendem a essa condição

random_integers[is_equal_to_3]

array([3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3])

Esses são os aspectos básicos dos arrays booleanos, que são úteis em muitas situações. Especificamente, você pode usar o método **.loc** de DataFrames para indexar suas linhas por uma máscara booleana e as colunas por rótulo. 

# Continuando a verificação da integridade dos dados

Com o conhecimento dos arrays booleanos, examinaremos alguns dos IDs duplicados que descobrimos.

Já verificamos que nenhum ID aparece mais de duas vezes. Podemos usar essa informação para localizar os IDs duplicados e examiná-los. Em seguida, tomaremos medidas para remover linhas de qualidade duvidosa do dataset.

In [21]:
# criando máscara booleana para condições iguais a 3

dupe_mask = id_counts == 2

In [22]:
# obtendo os cinco primeiros resutados

dupe_mask[:5]

6ff744d5-91f0    True
4653bc48-0141    True
5deff6b6-62ff    True
edd737fd-5b07    True
f0723a52-7018    True
Name: ID, dtype: bool

Aqui, **dupe_mask** é a máscara lógica que criamos para armazenar os valores booleanos.

In [23]:
# acessando o índice de id_count e exibindo as cinco primeiras linhas como contexto usando o comando a seguir

id_counts.index[:5]

Index(['6ff744d5-91f0', '4653bc48-0141', '5deff6b6-62ff', 'edd737fd-5b07',
       'f0723a52-7018'],
      dtype='object')

In [24]:
# selecionando e armazenando os IDs duplicados em uma nova variável chamada dupe_ids 

dupe_ids = id_counts.index[dupe_mask]

In [25]:
#exibindo os 15 primeiros ids armazenados

dupe_ids[:15]

Index(['6ff744d5-91f0', '4653bc48-0141', '5deff6b6-62ff', 'edd737fd-5b07',
       'f0723a52-7018', 'e77f7765-8731', '3d832783-1571', '15edb406-305b',
       '994330b6-eeec', '369e565a-26a1', '87dec940-75b7', 'e50d8395-da32',
       '04c32736-903c', 'e668604e-21f6', '4f249cbc-5e9c'],
      dtype='object')

In [26]:
# convertendo dupe_ids em uma lista e obtendo seu tamanho

dupe_ids = list(dupe_ids)
len(dupe_ids)

313

Alteramos a variável **dupe_ids** para uma **lista**, já que precisamos dela nessa forma em etapas futuras. A lista tem um tamanho igual a 313, como pode ser visto na saída acima, que coincide com o número de IDs duplicados que obtivemos na contagem de valores algumas linhas acima.

In [27]:
# verificando os dados de dupe_ids exibindo as cinco primeiras entradas

dupe_ids[:5]

['6ff744d5-91f0',
 '4653bc48-0141',
 '5deff6b6-62ff',
 'edd737fd-5b07',
 'f0723a52-7018']

Agora estamos prontos para examinar os dados dos IDs de nossa lista de duplicatas. Especificamente, queremos examinar os valores das características para ver se há algode diferente entre essas entradas.

Para isso, usaremos os métodos **.isin** e **.loc**.

Usando os três primeiros IDs de nossa lista de duplicatas, **dupe_ids[:3]**, primeiro queremos encontrar as linhas que contêm esses IDs. Se passarmos essa lista de IDs para o método **.isin** da série ID, ele criará outra máscara lógica que poderemos usar no DataFrame maior para exibir as linhas que contêm os IDs.

O método **.isin** ficará aninhado em uma instrução **.loc** de indexação do DataFrame para a seleção do local de todas as linhas que contêm 'True' na máscara booleana.

O segundo argumento da instrução de indexação **.loc** é :, que implica que todas as colunas serão selecionadas. Ao executar as próximas etapas, estaremos basicamente filtrando o DataFrame para visualizar todas as colunas dos três primeiros IDs duplicados.

In [28]:
# executando o comando formulado anteriormente

df.loc[df['ID'].isin(dupe_ids[:3]),:].head(10)

ID  LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE PAY_1  PAY_2  \
2678   4653bc48-0141     200000    2          2         2   43     0      0   
2778   4653bc48-0141          0    0          0         0    0     0      0   
3018   6ff744d5-91f0      20000    1          1         1   34     2      0   
3118   6ff744d5-91f0          0    0          0         0    0     0      0   
13146  5deff6b6-62ff     140000    2          2         2   32    -1     -1   
13246  5deff6b6-62ff          0    0          0         0    0     0      0   

       PAY_3  PAY_4  ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2  \
2678       0      0  ...     190775     196449     192085      8155      8302   
2778       0      0  ...          0          0          0         0         0   
3018       0      2  ...       8436      12016      11404      2000      2000   
3118       0      0  ...          0          0          0         0         0   
13146     -1     -2  ...          0       1186          0      3134         0   
13246      0      0  ...          0          0          0         0         0   

       PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  default payment next month  
2678       7225     10153      6987      5262                           0  
2778          0         0         0         0                           0  
3018          0      4000         0      4000                           0  
3118          0         0         0         0                           0  
13146         0      1186         0      1047                           1  
13246         0         0         0         0                           0  

[6 rows x 25 columns]

O que podemos ver aqui é que cada ID duplicado para ter uma linha de dados válidos e outra somente com zeros. Fica claro que devemos excluir as linhas somentes com zeros, já que não faz sentido alguém ter 0 anos, limite de crédito igual a 0 e assim por diante.

Uma abordagem para lidarmos com esse problema seria encontrar as linhas que só tem zeros, exceto na primeira coluna, que tem os IDs. Esses dados seriam inválidos e, se excluíssemos, também resolveríamos nosso problema de IDs duplicados.


In [29]:
# criando uma máscara booleana feature_zero_mask

df_zero_mask = df == 0

Usaremos **df_zero_mask**, que é outro DataFrame, contendo valores booleanos. O objetivo é criarmos uma série booleana, **feature_zero_mask**, que identifique cada linha em que todos os elementos a partir da segunda coluna (as características e a resposta, mas não o IDs) sejam 0.

Para isso, primeiro temos que indexar **df_zero_mask** usando o método de indexação de inteiros (.loc). Nesse método, passaremos (:) para examinar todas as linhas e (1:) para examinar todas as colunas a partir da segunda (ídice 1). Para concluir, aplicaremos o método **all()** ao longo do eixo da coluna (axis=1) e ele retornará **True** somente se todas as colunas dessa linha forem iguais a **True**. 

In [31]:
# criando a série booleana feature_zero_mask

feature_zero_mask = df_zero_mask.iloc[:,1:].all(axis=1)

In [32]:
# calculando a soma da série booleana

sum(feature_zero_mask)

315

Esse número é maior do que o número de IDs duplicados, logo, se excluirmos todas as "linhas de zeros", podemos nos livrar do problema dos IDs duplicados